lib/charms/opensearch/v0/models.py

```diff
@@ -346,3 +358,76 @@ def promote_failover(self) -> None:
         self.main_app = self.failover_app
         self.main_rel_id = self.failover_rel_id
         self.delete("failover")
+
+
+class OpenSearchPerfProfile(Model):
+    """Generates an immutable description of the performance profile."""
+
+    class Config:
+        """Pydantic config for this model."""
+
+        arbitrary_types_allowed = True
+
+    typ: PerformanceType
+    heap_size_in_kb: int = MIN_HEAP_SIZE
+    opensearch_yml: Dict[str, str] = {}
+    charmed_index_template: Dict[str, str] = {}
+    charmed_component_templates: Dict[str, str] = {}
+
+    @classmethod
+    def from_str(cls, input_str: str):
+        """Create a new instance of this class from a stringified json/dict repr."""
+        return cls(typ=PerformanceType(input_str))
+
+    @root_validator
+    def set_options(cls, values):  # noqa: N805
+        """Generate the attributes depending on the input."""
+        val = values["typ"]
+        if isinstance(val, str):
+            val = PerformanceType(val)
+
+        if val == PerformanceType.PRODUCTION:
+            values["heap_size_in_kb"] = min(
+                int(0.50 * OpenSearchPerfProfile.meminfo()["MemTotal"]),
+                MAX_HEAP_SIZE,
+            )
+
+        if val == PerformanceType.STAGING:
+            values["heap_size_in_kb"] = min(
+                int(0.25 * OpenSearchPerfProfile.meminfo()["MemTotal"]),
+                MAX_HEAP_SIZE,
+            )
+
+        if val == PerformanceType.TESTING:
+            values["heap_size_in_kb"] = MIN_HEAP_SIZE
+
+        if val != PerformanceType.TESTING:
+            values["opensearch_yml"] = {"indices.memory.index_buffer_size": "25%"}
+
+            values["charmed_index_template"] = {
+                "charmed-index-tpl": {
+                    "index_patterns": ["*"],
+                    "template": {
+                        "settings": {
+                            "number_of_replicas": "2",
+                        },
+                    },
+                },
+            }
+
+        return values
+
+    @staticmethod
+    def meminfo() -> dict[str, float]:
+        """Read the /proc/meminfo file and return the values.
+
+        According to the kernel source code, the values are always in kB:
+            https://github.com/torvalds/linux/blob/
+                2a130b7e1fcdd83633c4aa70998c314d7c38b476/fs/proc/meminfo.c#L31
+        """
+        with open("/proc/meminfo") as f:
+            meminfo = f.read()
```

Comment:
nit - maybe for clarity as the next lines are a bit difficult to read:
```suggestion
            meminfo = f.read().split("\n")
            meminfo = [line.split() for line in meminfo if line.strip()]
```